eventuals/tcp.h

```diff
@@ -0,0 +1,1455 @@
+  }
+
+  auto ReceiveSome(
+      void* destination,
+      size_t destination_size) {
+    return RescheduleAfter(
+        _Receive::Composable{
+            *this,
+            destination,
+            destination_size});
+  }
+
+  auto Receive(
+      void* destination,
+      size_t destination_size,
+      size_t bytes_to_read) {
+    return RescheduleAfter(
+        _ReceiveNBytes::Composable{
+            *this,
+            destination,
+            destination_size,
+            bytes_to_read});
+  }
+
+  auto Send(const void* source, size_t source_size) {
+    return RescheduleAfter(_Send::Composable{*this, source, source_size});
+  }
+
+  auto Shutdown(ShutdownType shutdown_type) {
+    return RescheduleAfter(_Shutdown::Composable{*this, shutdown_type});
+  }
+
+  auto Close() {
+    return RescheduleAfter(_Close::Composable{*this});
+  }
+
+  bool IsOpen() {
+    return underlying_handle().is_open();
+  }
+
+  uint16_t BindedPort() {
+    return underlying_handle().local_endpoint().port();
+  }
+
+  std::string BindedIp() {
+    return underlying_handle().local_endpoint().address().to_string();
+  }
+
+ private:
+  asio::ip::tcp::socket socket_;
+  Protocol protocol_ = Protocol::UNKNOWN;
+
+  asio::ip::tcp::socket& underlying_handle() {
+    return socket_;
+  }
+
+  friend class Acceptor;
+
+  struct _Open {
+    template <typename K_>
+    struct Continuation {
+      Continuation(K_ k, Socket& socket, Protocol protocol)
+        : k_(std::move(k)),
+          socket_(socket),
+          protocol_(protocol) {}
+
+      Continuation(Continuation&& that)
+        : k_(std::move(that.k_)),
+          socket_(that.socket_),
+          protocol_(that.protocol_) {
+        CHECK(!that.started_ || !that.completed_) << "moving after starting";
+        CHECK(!handler_);
+      }
+
+      void Start() {
+        CHECK(!started_ && !completed_);
+
+        started_ = true;
+        completed_ = true;
+
+        asio::error_code ec;
+
+        switch (protocol_) {
+          case Protocol::IPV4:
+            socket_.underlying_handle().open(asio::ip::tcp::v4(), ec);
+            break;
+          case Protocol::IPV6:
+            socket_.underlying_handle().open(asio::ip::tcp::v6(), ec);
+            break;
+          default:
+            k_.Fail("Unknown protocol");
+            return;
+        }
+
+        socket_.protocol_ = protocol_;
+
+        if (!ec) {
+          k_.Start();
+        } else {
+          k_.Fail(ec.message());
+        }
+      }
+
+      template <typename... Args>
+      void Fail(Args&&... args) {
+        k_.Fail(std::forward<Args>(args)...);
+      }
+
+      void Stop() {
+        k_.Stop();
+      }
+
+      void Register(class Interrupt& interrupt) {
+        k_.Register(interrupt);
+
+        handler_.emplace(&interrupt, [this]() {
+          if (!completed_) {
+            completed_ = true;
+            k_.Stop();
+          }
+        });
```

Comment:
There's a race here with `Start()`. The reason why it's safe to follow this pattern with libuv is because we do `loop_.Submit()` which ensures that the interrupt handler and start will execute serially.

I also don't think that this interrupt handler really buys us very much, as in, there's nothing asynchronous about `open()`, and we don't need to support interrupting things that aren't asynchronous just haven't started yet.

I could see us saving a pointer to `interrupt` and then in `Start()` check to see if we've been interrupted and then do a `k.Stop()` instead of doing the open.

Looking at the rest of this file makes it seem like there are quite a few places where we don't have asynchronous code that we can just replace with an `Eventual()`, for example:

```cpp
class Socket {
 public:
  auto Open() {
    return Eventual<void>()
      .interruptible()
      .start([](auto& k, Interrupt::Handler& handler) {
        if (handler.interrupt().Triggered()) {
          k.Stop();
        } else {
          ...
          socket_.open(..., error);
          if (!error) {
            k.Start();
          } else {
            k.Fail(..);
          }
          ...
        }
      });
  }
};
```

The actual asynchronous functions, like the ones which do `async_read()` or `async_write()` can also follow this pattern because if I understand correctly from reading the documentation as well as from seeing the code in this PR we don't have any way of interrupting them except for stopping the `io_context`, which doesn't seem like something we can do. So, if they aren't interruptible, we might as well follow the simpler pattern.

Reply:
.cancel() cancels all asynchronous operations associated with the acceptor/socket.

Reply:
Ok great! Is it safe to be called from any thread?

If the answer is yes, then I still like the strategy suggested by instead of just checking if the interrupt has been triggered we can add a handler which will call `cancel()`!

Reply:
That's a great question, I do not see it being answered explicitly by documentation, I'll do some testing on it.